File: pyAmpliCol/src/pyamplicol/api_templates/cpp/check_standalone.cpp
#include <rusticol.hpp>

#include <cmath>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace fs = std::filesystem;

struct Options {
    std::string process;
    std::string model_parameters;
    std::vector<std::string> parameter_names;
    std::vector<double> parameter_real;
    std::vector<double> parameter_imaginary;
    int precision = 16;
    bool json = false;
};

std::string json_string(const std::string &value) {
    std::ostringstream output;
    output << '"';
    for (const unsigned char character : value) {
        switch (character) {
        case '"': output << "\\\""; break;
        case '\\': output << "\\\\"; break;
        case '\n': output << "\\n"; break;
        case '\r': output << "\\r"; break;
        case '\t': output << "\\t"; break;
        default:
            if (character < 0x20) {
                output << "\\u" << std::hex << std::setw(4) << std::setfill('0')
                       << static_cast<int>(character) << std::dec << std::setfill(' ');
            } else {
                output << character;
            }
        }
    }
    output << '"';
    return output.str();
}

Options parse_options(int argc, char **argv) {
    Options options;
    for (int index = 1; index < argc; ++index) {
        const std::string argument = argv[index];
        auto require = [&](int count) {
            if (index + count >= argc) {
                throw std::invalid_argument("missing value after " + argument);
            }
        };
        if (argument == "--process") {
            require(1);
            options.process = argv[++index];
        } else if (argument == "--model-parameters") {
            require(1);
            options.model_parameters = argv[++index];
        } else if (argument == "--set-parameter") {
            require(3);
            options.parameter_names.emplace_back(argv[++index]);
            options.parameter_real.push_back(std::stod(argv[++index]));
            options.parameter_imaginary.push_back(std::stod(argv[++index]));
        } else if (argument == "--precision") {
            require(1);
            options.precision = std::stoi(argv[++index]);
        } else if (argument == "--json") {
            options.json = true;
        } else if (argument == "--help" || argument == "-h") {
            std::cout << "usage: check_standalone [--process KEY] [--model-parameters PATH] "
                         "[--set-parameter NAME REAL IMAG] [--precision 16] [--json]\n";
            std::exit(0);
        } else {
            throw std::invalid_argument("unknown option: " + argument);
        }
    }
    if (options.precision != 16) {
        throw std::invalid_argument(
            "the C++ Rusticol API supports only double precision (--precision 16)");
    }
    return options;
}

std::vector<std::string> split_tabs(const std::string &line) {
    std::vector<std::string> fields;
    std::size_t start = 0;
    while (true) {
        const auto separator = line.find('\t', start);
        fields.push_back(line.substr(start, separator - start));
        if (separator == std::string::npos) break;
        start = separator + 1;
    }
    return fields;
}

std::vector<double> load_validation_point(
    const fs::path &path, const std::string &process_key, std::size_t external_count) {
    std::ifstream input(path);
    if (!input) return {};
    std::string line;
    std::getline(input, line);
    if (line != "RUSTICOL_VALIDATION_POINTS_V1") {
        throw std::runtime_error("unsupported validation_points.dat format");
    }
    while (std::getline(input, line)) {
        if (line.empty() || line[0] == '#') continue;
        const auto fields = split_tabs(line);
        if (fields.size() < 2 || fields[0] != process_key) continue;
        const auto row_external_count = static_cast<std::size_t>(std::stoull(fields[1]));
        if (row_external_count != external_count || fields.size() != 2 + 4 * external_count) {
            throw std::runtime_error("validation point has an incompatible external-particle count");
        }
        std::vector<double> values;
        values.reserve(4 * external_count);
        for (std::size_t index = 2; index < fields.size(); ++index) {
            values.push_back(std::stod(fields[index]));
        }
        return values;
    }
    return {};
}

void write_int_array(std::ostream &output, const std::vector<std::int32_t> &values) {
    output << '[';
    for (std::size_t index = 0; index < values.size(); ++index) {
        if (index) output << ',';
        output << values[index];
    }
    output << ']';
}

void write_size_array(std::ostream &output, const std::vector<std::size_t> &values) {
    output << '[';
    for (std::size_t index = 0; index < values.size(); ++index) {
        if (index) output << ',';
        output << values[index];
    }
    output << ']';
}

void write_common_json(
    std::ostream &output,
    const rusticol::Runtime &runtime,
    const std::vector<rusticol::ExternalParticle> &particles,
    const std::vector<rusticol::HelicityConfiguration> &helicities,
    const std::vector<rusticol::ColorComponent> &colors) {
    output << "\"process\":" << json_string(runtime.process())
           << ",\"process_key\":" << json_string(runtime.process_key())
           << ",\"color_accuracy\":" << json_string(runtime.color_accuracy());
    output << ",\"external_particles\":[";
    for (std::size_t index = 0; index < particles.size(); ++index) {
        if (index) output << ',';
        output << "{\"index\":" << particles[index].index
               << ",\"pdg\":" << particles[index].pdg << '}';
    }
    output << "],\"helicities\":[";
    for (std::size_t index = 0; index < helicities.size(); ++index) {
        if (index) output << ',';
        output << "{\"id\":" << json_string(helicities[index].id)
               << ",\"helicities\":";
        write_int_array(output, helicities[index].helicities);
        output << '}';
    }
    output << "],\"colors\":[";
    for (std::size_t index = 0; index < colors.size(); ++index) {
        if (index) output << ',';
        output << "{\"id\":" << json_string(colors[index].id)
               << ",\"kind\":" << json_string(colors[index].kind)
               << ",\"word\":";
        write_size_array(output, colors[index].word);
        output << '}';
    }
    output << ']';
}

int main(int argc, char **argv) {
    try {
        const auto options = parse_options(argc, argv);
        const fs::path executable = fs::weakly_canonical(fs::absolute(argv[0]));
        const fs::path root = executable.parent_path().parent_path().parent_path();
        rusticol::Runtime runtime(root.string(), options.process, options.model_parameters);
        if (!options.parameter_names.empty()) {
            runtime.set_model_parameters(
                options.parameter_names, options.parameter_real, options.parameter_imaginary);
        }
        const auto particles = runtime.external_particles();
        const auto helicities = runtime.helicities();
        const auto colors = runtime.colors();
        const auto momenta = load_validation_point(
            root / "API" / "validation_points.dat", runtime.process_key(), particles.size());
        if (momenta.empty()) {
            if (options.json) {
                std::cout << "{\"language\":\"cpp\",\"available\":false,";
                write_common_json(std::cout, runtime, particles, helicities, colors);
                std::cout << ",\"diagnostic\":\"no bundled validation point is available\"}\n";
            } else {
                std::cout << "process: " << runtime.process() << "\n"
                          << "no bundled validation point is available; metadata load succeeded\n";
            }
            return 0;
        }
        const auto total = runtime.evaluate(momenta, 1);
        const auto resolved = runtime.evaluate_resolved(momenta, 1);
        const auto explicit_total = resolved.total();

        if (options.json) {
            std::cout << std::setprecision(17)
                      << "{\"language\":\"cpp\",\"available\":true,\"precision\":16,";
            write_common_json(std::cout, runtime, particles, helicities, colors);
            std::cout << ",\"shape\":[1," << resolved.helicities.size() << ','
                      << resolved.colors.size() << "],\"values\":[";
            for (std::size_t index = 0; index < resolved.values.size(); ++index) {
                if (index) std::cout << ',';
                std::cout << resolved.values[index];
            }
            std::cout << "],\"resolved_sum\":[" << explicit_total[0]
                      << "],\"compatibility_total\":[" << total[0] << "]}\n";
        } else {
            std::cout << std::setprecision(17)
                      << "process: " << runtime.process() << " [" << runtime.process_key() << "]\n"
                      << "resolved shape: (1, " << resolved.helicities.size() << ", "
                      << resolved.colors.size() << ")\n";
            for (std::size_t helicity = 0; helicity < resolved.helicities.size(); ++helicity) {
                for (std::size_t color = 0; color < resolved.colors.size(); ++color) {
                    std::cout << "  " << resolved.helicities[helicity].id << "  "
                              << resolved.colors[color].id << "  "
                              << resolved(0, helicity, color) << '\n';
                }
            }
            std::cout << "explicit resolved sum: " << explicit_total[0] << '\n'
                      << "compatibility total:   " << total[0] << '\n';
        }
        const double scale = std::max(std::abs(total[0]), 1.0);
        if (std::abs(explicit_total[0] - total[0]) > 1.0e-12 * scale) {
            throw std::runtime_error(
                "resolved components do not reproduce the compatibility total");
        }
        return 0;
    } catch (const std::exception &error) {
        std::cerr << "check_standalone: " << error.what() << '\n';
        return 1;
    }
}

File: pyAmpliCol/src/pyamplicol/api_templates/fortran/check_standalone.f90
program check_standalone
  use, intrinsic :: iso_c_binding
  use, intrinsic :: iso_fortran_env, only: error_unit
  use rusticol
  implicit none

  integer, parameter :: MAX_OVERRIDES = 128
  type(rusticol_runtime) :: runtime
  type(rusticol_external_particle), allocatable :: particles(:)
  type(rusticol_helicity_configuration), allocatable :: helicities(:)
  type(rusticol_color_component), allocatable :: colors(:)
  character(len=4096) :: process_dir, process_key, model_parameters
  character(len=4096) :: executable
  character(len=256) :: override_names_buffer(MAX_OVERRIDES)
  character(len=256), allocatable :: override_names(:)
  real(c_double) :: override_real_buffer(MAX_OVERRIDES)
  real(c_double) :: override_imaginary_buffer(MAX_OVERRIDES)
  real(c_double), allocatable :: override_real(:), override_imaginary(:)
  real(c_double), allocatable, target :: momenta(:), totals(:), resolved(:, :, :)
  logical :: json_output, point_available
  integer :: precision, override_count
  integer(c_int) :: status
  integer :: color_index, helicity_index
  real(c_double) :: explicit_total, scale

  call parse_options(process_key, model_parameters, precision, json_output, &
      override_names_buffer, override_real_buffer, override_imaginary_buffer, override_count)
  if (precision /= 16) call fail( &
      "the Fortran Rusticol API supports only double precision (--precision 16)")
  call resolve_process_dir(executable, process_dir)

  if (len_trim(process_key) > 0 .and. len_trim(model_parameters) > 0) then
    call runtime%load(trim(process_dir), trim(process_key), trim(model_parameters), status)
  else if (len_trim(process_key) > 0) then
    call runtime%load(trim(process_dir), trim(process_key), ierr=status)
  else if (len_trim(model_parameters) > 0) then
    call runtime%load(trim(process_dir), model_parameters=trim(model_parameters), ierr=status)
  else
    call runtime%load(trim(process_dir), ierr=status)
  end if
  if (status /= RUSTICOL_STATUS_OK) call fail(rusticol_last_error())

  if (override_count > 0) then
    allocate(override_names(override_count), override_real(override_count), &
        override_imaginary(override_count))
    override_names = override_names_buffer(1:override_count)
    override_real = override_real_buffer(1:override_count)
    override_imaginary = override_imaginary_buffer(1:override_count)
    call runtime%set_model_parameters( &
        override_names, override_real, override_imaginary, ierr=status)
    if (status /= RUSTICOL_STATUS_OK) call fail(rusticol_last_error())
  end if

  particles = runtime%external_particles(ierr=status)
  if (status /= RUSTICOL_STATUS_OK) call fail(rusticol_last_error())
  helicities = runtime%helicities(ierr=status)
  if (status /= RUSTICOL_STATUS_OK) call fail(rusticol_last_error())
  colors = runtime%colors(ierr=status)
  if (status /= RUSTICOL_STATUS_OK) call fail(rusticol_last_error())
  process_key = runtime%process_key(ierr=status)
  if (status /= RUSTICOL_STATUS_OK) call fail(rusticol_last_error())
  call load_validation_point( &
      trim(process_dir) // "/API/validation_points.dat", trim(process_key), &
      size(particles), momenta, point_available)

  if (.not. point_available) then
    if (json_output) then
      call emit_json_prefix(runtime, particles, helicities, colors, .false.)
      write(*, '(A)') ',"diagnostic":"no bundled validation point is available"}'
    else
      write(*, '(A)') "process: " // runtime%process()
      write(*, '(A)') "no bundled validation point is available; metadata load succeeded"
    end if
    stop 0
  end if

  call runtime%evaluate(momenta, 1_c_size_t, totals, ierr=status)
  if (status /= RUSTICOL_STATUS_OK) call fail(rusticol_last_error())
  call runtime%evaluate_resolved(momenta, 1_c_size_t, resolved, ierr=status)
  if (status /= RUSTICOL_STATUS_OK) call fail(rusticol_last_error())
  explicit_total = sum(resolved(:, :, 1))

  if (json_output) then
    call emit_json_prefix(runtime, particles, helicities, colors, .true.)
    write(*, '(A,I0,A,I0,A)', advance='no') ',"shape":[1,', size(resolved, 2), ',', &
        size(resolved, 1), '],"values":['
    do helicity_index = 1, size(resolved, 2)
      do color_index = 1, size(resolved, 1)
        if (helicity_index > 1 .or. color_index > 1) write(*, '(A)', advance='no') ','
        write(*, '(G0.17)', advance='no') resolved(color_index, helicity_index, 1)
      end do
    end do
    write(*, '(A,G0.17,A,G0.17,A)') '],"resolved_sum":[', explicit_total, &
        '],"compatibility_total":[', totals(1), ']}'
  else
    write(*, '(A)') "process: " // runtime%process() // " [" // trim(process_key) // "]"
    write(*, '(A,I0,A,I0,A)') "resolved shape: (1, ", size(resolved, 2), ", ", &
        size(resolved, 1), ")"
    do helicity_index = 1, size(helicities)
      do color_index = 1, size(colors)
        write(*, '(2X,A,2X,A,2X,G0.17)') trim(helicities(helicity_index)%id), &
            trim(colors(color_index)%id), resolved(color_index, helicity_index, 1)
      end do
    end do
    write(*, '(A,G0.17)') "explicit resolved sum: ", explicit_total
    write(*, '(A,G0.17)') "compatibility total:   ", totals(1)
  end if
  scale = max(abs(totals(1)), 1.0_c_double)
  if (abs(explicit_total - totals(1)) > 1.0e-12_c_double * scale) then
    call fail("resolved components do not reproduce the compatibility total")
  end if

contains

  subroutine fail(message)
    character(len=*), intent(in) :: message
    write(error_unit, '(A)') "check_standalone: " // trim(message)
    stop 1
  end subroutine fail

  subroutine parse_options(selected_process, model_card, selected_precision, json, &
      names, real_values, imaginary_values, count)
    character(len=*), intent(out) :: selected_process, model_card
    integer, intent(out) :: selected_precision, count
    logical, intent(out) :: json
    character(len=*), intent(out) :: names(:)
    real(c_double), intent(out) :: real_values(:), imaginary_values(:)
    character(len=4096) :: argument, value
    integer :: argc, index, parse_status

    selected_process = ""
    model_card = ""
    selected_precision = 16
    count = 0
    json = .false.
    argc = command_argument_count()
    index = 1
    do while (index <= argc)
      call get_command_argument(index, argument)
      select case (trim(argument))
      case ("--process")
        if (index + 1 > argc) call fail("missing value after --process")
        index = index + 1
        call get_command_argument(index, selected_process)
      case ("--model-parameters")
        if (index + 1 > argc) call fail("missing value after --model-parameters")
        index = index + 1
        call get_command_argument(index, model_card)
      case ("--set-parameter")
        if (index + 3 > argc) call fail("--set-parameter requires NAME REAL IMAG")
        if (count >= size(names)) call fail("too many --set-parameter options")
        count = count + 1
        index = index + 1
        call get_command_argument(index, names(count))
        index = index + 1
        call get_command_argument(index, value)
        read(value, *, iostat=parse_status) real_values(count)
        if (parse_status /= 0) call fail("invalid real model-parameter component")
        index = index + 1
        call get_command_argument(index, value)
        read(value, *, iostat=parse_status) imaginary_values(count)
        if (parse_status /= 0) call fail("invalid imaginary model-parameter component")
      case ("--precision")
        if (index + 1 > argc) call fail("missing value after --precision")
        index = index + 1
        call get_command_argument(index, value)
        read(value, *, iostat=parse_status) selected_precision
        if (parse_status /= 0) call fail("invalid --precision value")
      case ("--json")
        json = .true.
      case ("--help", "-h")
        write(*, '(A)') "usage: check_standalone [--process KEY] " // &
            "[--model-parameters PATH] [--set-parameter NAME REAL IMAG] " // &
            "[--precision 16] [--json]"
        stop 0
      case default
        call fail("unknown option: " // trim(argument))
      end select
      index = index + 1
    end do
  end subroutine parse_options

  subroutine resolve_process_dir(program_path, root)
    character(len=*), intent(out) :: program_path, root
    integer :: slash
    call get_command_argument(0, program_path)
    slash = scan(trim(program_path), "/", back=.true.)
    if (slash == 0) then
      root = "../.."
    else
      root = program_path(:slash - 1) // "/../.."
    end if
  end subroutine resolve_process_dir

  subroutine load_validation_point(path, key, external_count, values, available)
    character(len=*), intent(in) :: path, key
    integer, intent(in) :: external_count
    real(c_double), allocatable, target, intent(out) :: values(:)
    logical, intent(out) :: available
    character(len=65536) :: line
    character(len=4096) :: row_key
    integer :: unit, io_status, row_count

    available = .false.
    allocate(values(4 * external_count))
    open(newunit=unit, file=path, status="old", action="read", iostat=io_status)
    if (io_status /= 0) return
    read(unit, '(A)', iostat=io_status) line
    if (io_status /= 0 .or. trim(line) /= "RUSTICOL_VALIDATION_POINTS_V1") then
      close(unit)
      call fail("unsupported validation_points.dat format")
    end if
    do
      read(unit, '(A)', iostat=io_status) line
      if (io_status /= 0) exit
      if (len_trim(line) == 0 .or. line(1:1) == "#") cycle
      read(line, *, iostat=io_status) row_key, row_count
      if (io_status /= 0 .or. trim(row_key) /= trim(key)) cycle
      if (row_count /= external_count) then
        close(unit)
        call fail("validation point has an incompatible external-particle count")
      end if
      read(line, *, iostat=io_status) row_key, row_count, values
      if (io_status /= 0) then
        close(unit)
        call fail("invalid validation point row")
      end if
      available = .true.
      exit
    end do
    close(unit)
  end subroutine load_validation_point

  function json_escape(value) result(escaped)
    character(len=*), intent(in) :: value
    character(len=:), allocatable :: escaped
    integer :: index
    escaped = ""
    do index = 1, len_trim(value)
      select case (value(index:index))
      case ('"')
        escaped = escaped // '\\"'
      case ('\\')
        escaped = escaped // '\\\\'
      case default
        escaped = escaped // value(index:index)
      end select
    end do
  end function json_escape

  subroutine emit_json_prefix(active_runtime, external, helicity_items, color_items, available)
    type(rusticol_runtime), intent(inout) :: active_runtime
    type(rusticol_external_particle), intent(in) :: external(:)
    type(rusticol_helicity_configuration), intent(in) :: helicity_items(:)
    type(rusticol_color_component), intent(in) :: color_items(:)
    logical, intent(in) :: available
    integer :: item, component

    write(*, '(A)', advance='no') '{"language":"fortran","available":'
    if (available) then
      write(*, '(A)', advance='no') 'true,"precision":16'
    else
      write(*, '(A)', advance='no') 'false'
    end if
    write(*, '(A)', advance='no') ',"process":"' // &
        json_escape(active_runtime%process()) // '","process_key":"' // &
        json_escape(active_runtime%process_key()) // '","color_accuracy":"' // &
        json_escape(active_runtime%color_accuracy()) // '","external_particles":['
    do item = 1, size(external)
      if (item > 1) write(*, '(A)', advance='no') ','
      write(*, '(A,I0,A,I0,A)', advance='no') '{"index":', external(item)%index, &
          ',"pdg":', external(item)%pdg, '}'
    end do
    write(*, '(A)', advance='no') '],"helicities":['
    do item = 1, size(helicity_items)
      if (item > 1) write(*, '(A)', advance='no') ','
      write(*, '(A)', advance='no') '{"id":"' // &
          json_escape(helicity_items(item)%id) // '","helicities":['
      do component = 1, size(helicity_items(item)%helicities)
        if (component > 1) write(*, '(A)', advance='no') ','
        write(*, '(I0)', advance='no') helicity_items(item)%helicities(component)
      end do
      write(*, '(A)', advance='no') ']}'
    end do
    write(*, '(A)', advance='no') '],"colors":['
    do item = 1, size(color_items)
      if (item > 1) write(*, '(A)', advance='no') ','
      write(*, '(A)', advance='no') '{"id":"' // json_escape(color_items(item)%id) // &
          '","kind":"' // json_escape(color_items(item)%kind) // '","word":['
      do component = 1, size(color_items(item)%word)
        if (component > 1) write(*, '(A)', advance='no') ','
        write(*, '(I0)', advance='no') color_items(item)%word(component)
      end do
      write(*, '(A)', advance='no') ']}'
    end do
    write(*, '(A)', advance='no') ']'
  end subroutine emit_json_prefix

end program check_standalone
